library/alloc/src/string.rs

```diff
@@ -843,6 +843,42 @@ impl String {
         self.vec.extend_from_slice(string.as_bytes())
     }
 
+    /// Copies elements from `src` range to the end of the string.
+    ///
+    /// ## Panics
+    ///
+    /// Panics if the starting point or end point do not lie on a [`char`]
+    /// boundary, or if they're out of bounds.
+    ///
+    /// ## Examples
+    ///
+    /// ```
+    /// #![feature(string_extend_from_within)]
+    /// let mut string = String::from("abcde");
+    ///
+    /// string.extend_from_within(2..);
+    /// assert_eq!(string, "abcdecde");
+    ///
+    /// string.extend_from_within(..2);
+    /// assert_eq!(string, "abcdecdeab");
+    ///
+    /// string.extend_from_within(4..8);
+    /// assert_eq!(string, "abcdecdeabecde");
+    /// ```
+    #[cfg(not(no_global_oom_handling))]
+    #[unstable(feature = "string_extend_from_within", issue = "none")]
+    pub fn extend_from_within<R>(&mut self, src: R)
+    where
+        R: RangeBounds<usize>,
+    {
+        let src @ Range { start, end } = slice::range(src, ..self.len());
+
+        assert!(self.is_char_boundary(start));
+        assert!(self.is_char_boundary(end));
```

Comment:
If panic happens, how does one know if it panics because of start or end?

Reply:
I think it should be possible by checking the line number in the panic message. It isn't convenient, but it is possible.

Also, note that the same pattern appears multiple times in this file: https://github.com/rust-lang/rust/blob/23f9b92c5e3fd6dcff19fb185b64990f0336a8bf/library/alloc/src/string.rs#L1643-L1644 https://github.com/rust-lang/rust/blob/23f9b92c5e3fd6dcff19fb185b64990f0336a8bf/library/alloc/src/string.rs#L1688-L1700

Reply:
Maybe we should have a separate issue for this.

Reply:
Wait, it's not necessary. The current panic error message is good enough.
```
thread 'main' panicked at 'assertion failed: self.is_char_boundary(start)', /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/alloc/src/string.rs:1571:9
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4e29eec2c640b0a42b054f58b1f6753d

Reply:
Ah, so when called without a message, `assert!` prints the passed expression. That's nice.